### PR TITLE
Feature curved tetrahedron

### PR DESCRIPTION
**Pull request by jfussbro**
**_Describe your changes here:_**

Enables usability of curved tetrahedral elements. Also fixes a bug in the readmshfile.

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually

- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [ ] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]
Here are some first comments :)

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.hxx [hunk 1]
```diff
@@ -304,7 +304,18 @@ struct t8_geometry_occ: public t8_geometry_with_vertices
                              double *out_coords) const;
 
   /**
-   * Maps points in the reference space \f$ \f$ [0,1]^3 \f$ \f$ to \f$ \mathbb{R}^3 \f$. Only for hex trees.
+   * Map a point in the reference space $$[0,1]^3$$ to $$\mathbb R^3$$. Only for tet trees.
+   * \param [in]  cmesh      The cmesh in which the point lies.
+   * \param [in]  gtreeid    The global tree (of the cmesh) in which the reference point is.
+   * \param [in]  ref_coords  Array of 3 entries, specifying a point in [0,1]^3.
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.hxx:310]:
```suggestion
   * \param [in]  ref_coords  Array of \a dimension x \a num_coords many entries, specifying points in \f$ [0,1]^\mathrm{dim} \f$.
   * \param [in]  num_coords  The number of points to map.
```

File: src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.cxx [hunk 2]
```diff
@@ -101,6 +105,9 @@ t8_geometry_occ::t8_geom_evaluate (t8_cmesh_t cmesh, t8_gloidx_t gtreeid, const
   case T8_ECLASS_HEX:
     t8_geometry_occ::t8_geom_evaluate_occ_hex (cmesh, gtreeid, ref_coords, 1, out_coords);
     break;
+  case T8_ECLASS_TET:
+    t8_geometry_occ::t8_geom_evaluate_occ_tet (cmesh, gtreeid, ref_coords, out_coords);
```
Comment (jmark) [inline src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.cxx:109]:
```suggestion
    t8_geometry_occ::t8_geom_evaluate_occ_tet (cmesh, gtreeid, ref_coords, 1, out_coords);
```
See comments below.

**Review by jmark — CHANGES_REQUESTED** [review]
Nice work! I like your code. It is well commented and clean & lean.

As discussed, please also add triangle/tetraeder support for the Curved Mesh Tutorial.

**Review by jmark — CHANGES_REQUESTED** [review]

**Review by jmark — APPROVED** [review]

**Comment by jmark** [comment]
LGTM! @sandro-elsweijer If you agree we can merge.

**Comment by sandro-elsweijer** [comment]
I think it would be best to wait for this, till the evaluations are completed. The usage of the advection solver is good for testing the code and some bugs could become observable :)

**Comment by jmark** [comment]
@jfussbro @sandro-elsweijer 

I encountered a strange phenomenon when I computed interpolation nodes with curved triangles. I present you a (minimal) working example via a curved mesh around a cylinder. The respective files to reproduce the results are attached to this comment. 

But first let me show what the problem is:
grafik
This is a scatter plots of interpolation nodes computed via the Julia interface T8code.jl which in turn is linked against the t8code library with the curved tri/tets merged in. I shifted one curved triangle into empty space on purpose to better show the details. The colors of each point represent the Jacobians at this point in space computed with the polynomial spanned by the interpolation nodes. Clearly the curved elements have some troubles. Even intersecting into neighboring elements. The refinement level is set to 0 in this case.

I did the same with a C++ version (attached to this comment) and got the same strange behavior:
grafik

To reproduce this please consider these files:
[t8_features_curved_triangles_generate_cmesh_cylinder_2d.geo.txt](https://github.com/DLR-AMR/t8code/files/13505378/t8_features_curved_triangles_generate_cmesh_cylinder_2d.geo.txt)
[scatter_plot.cxx.txt](https://github.com/DLR-AMR/t8code/files/13505379/scatter_plot.cxx.txt)
[scatter_plot.gpi.txt](https://github.com/DLR-AMR/t8code/files/13505380/scatter_plot.gpi.txt)

**Note above C++ file is linked against t8code with curved tri/tets feature which in turn was merged with the current main. There was a change in the reference basis vectors for tris/tets which is not merged in this PR yet. The same erraneous behavior is observed with the old reference basis, too.**

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]

File: src/t8_geometry/t8_geometry_helpers.c [hunk 3]
```diff
@@ -495,4 +590,4 @@ t8_plane_point_inside (const double point_on_face[3], const double face_normal[3
     return 0;
   }
   return 1;
-}
\ No newline at end of file
+}
```
[the remark this answers is not in the document]
Comment (sandro-elsweijer) [reply src/t8_geometry/t8_geometry_helpers.c:586]: Newline at EOF is missing

[the remark this answers is not in the document]
Comment (jfussbro) [reply src/t8_geometry/t8_geometry_helpers.c:586]: The new line is deleted by the indention script. But right now there is one black line at the EOF. Isn't that enough?

**Review by sandro-elsweijer — APPROVED** [review]

**Pull request merged**